spark_connect_labextension/sparkconnectserver/cluster.py

```diff
@@ -0,0 +1,227 @@
+        Get the Spark Connect server port
+
+        :returns: port number
+        """
+        return SPARK_CONNECT_PORT
+
+    def get_options(self) -> dict:
+        """
+        Get the Spark options for running the Connect server
+
+        :returns: dict of options
+        """
+        return self.options
+
+    def is_connect_server_running(self) -> bool:
+        """
+        Determine if the Spark Connect server process is running
+
+        :returns: boolean indicating if the server is running
+        """
+        env_variables = self.get_envs()
+        run_script = f"sh $SPARK_HOME/sbin/spark-daemon.sh status org.apache.spark.sql.connect.service.SparkConnectServer 1"
+        if self.pre_script:
+            run_script = self.pre_script + ' && ' + run_script
+
+        retcode = subprocess.Popen(run_script, shell=True, env=env_variables).wait()
+        return retcode == 0
+
+    def is_server_ready(self) -> bool:
+        """
+        Determine if the Spark Connect server is ready and exposing the gRPC socket through the specified port.
+
+        :returns: boolean indicating if the server is ready
+        """
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        result = sock.connect_ex(('127.0.0.1', self.get_port()))
+        sock.close()
+        return result == 0
+
+    def get_envs(self, envs: dict = {}) -> dict:
+        """
+        Get the current environment variables appended with the supplied env variables
+
+        :param envs: dict of additional environment variables (EnvName=EnvValue)
+        :returns: dict of environment variables (EnvName=EnvValue)
+        """
+        my_env = os.environ.copy()
+        if envs:
+            for key in envs:
+                my_env[key] = envs[key]
+
+        if SPARK_HOME:
+            my_env['SPARK_HOME'] = SPARK_HOME
+
+        return my_env
+
+    def get_config_args(self, options: dict) -> str:
+        """
+        Translate the Spark options dict into command line arguments.
+
+        :param options: dict of Spark options (OptionName=OptionValue)
+        :returns: string of command line arguments
+        """
+        if not options:
+            return []
+
+        args = []
+        for key, val in options.items():
+            args.append("--conf")
+            val = self.replace_env_params(val)
+            val = val.replace('"', '\\"')
+            args.append(f'{key}="{val}"')
+        return ' '.join(args)
+
+    def replace_env_params(self, value):
+        """
+        Replace {ENV_NAME} placeholder with actual env value
+
+        :param value: formatted string
+        :returns: value with replaced {ENV_NAME}
+        """
+        value = f"{value}"
+        replacable_values = {}
+        for _, variable, _, _ in Formatter().parse(value):
+            if variable is not None:
+                replacable_values[variable] = os.getenv(variable, '')
+
+        return value.format(**replacable_values)
+
+    def __del__(self):
```

Comment:
There is no guarantee that `__del__` will be executed. The `cluster` object will most probably only be destroyed at tear down time, when this module disappears (and hence the last reference to `cluster`), but `__del__` might not be invoked. You should better register a function with `atexit`: https://docs.python.org/3/library/atexit.html